### DISCUSSION TITLE
Using @OrthotropicBehaviour<Pipe> and ComputeStiffnessTensor<>

### DISCUSSION THREAD

**Discussion by lmersel** [thelfer/MFrontGenericInterfaceSupport#103] [Q&A]
Hello,

I am using mfront to define an new energy decomposition for a phase fracture modeling, and for any anisotropic elastic materials (ref  [Nguyen et al.](https://hal.archives-ouvertes.fr/hal-02861347/document) ) . It is similar of what Jeremy Bleyer did with Amor and Miehe decomposition coupling with fenics. At the end,  I would like to compare the three formulation in dynamic modelling.

The first step of the law is to provide as an input an anisotropic ( or at least orthotopic) stiffness matrix to mfront (the variable C). Then I have to compute the eigenvalues and and eigenprojectors of C. At the end, I should transform the stiffness matrix into C^(1/2) and  C^(-1/2) like in the joined image. 
<img width="396" alt="decompositionStiffnessTensor" src="https://user-images.githubusercontent.com/110243126/195014476-9a8d8e0f-cd80-428f-9f31-0e4122b334c0.PNG">

The matrix C is considered constant so the decomposition should carry out only once. 
I started to code this step with 3 blocks : 
`@OrthotropicBehaviour<Pipe>;`
`@ComputeStiffnessTensor<UnAltered> {};`
`@InitLocalVariables{}`

At the compilation I got this  error : 
`In file included from PhaseFieldDisplacementNguyenSplit-generic.cxx:27:
PhaseFieldDisplacementNguyenSplit.mfront: In instantiation of 'void tfel::material::PhaseFieldDisplacementNguyenSplit<hypothesis, Numeric                                                                                Type, false>::initialize() [with tfel::material::ModellingHypothesis::Hypothesis hypothesis = tfel::material::ModellingHypothesis::AXISYM                                                                                METRICALGENERALISEDPLANESTRAIN; NumericType = double]':
/stck/lmersel/CODE_PROGRAMME/MFront_MGIS/TFEL-4.0/install/include/MFront/GenericBehaviour/Integrate.hxx:318:19:   required from 'int mfro                                                                                nt::gb::integrate(mfront_gb_BehaviourData&, typename Behaviour::SMFlag, tfel::material::OutOfBoundsPolicy) [with Behaviour = tfel::materi                                                                                al::PhaseFieldDisplacementNguyenSplit<tfel::material::ModellingHypothesis::AXISYMMETRICALGENERALISEDPLANESTRAIN, double, false>; mfront_g                                                                                b_BehaviourData = mfront_gb_BehaviourData; typename Behaviour::SMFlag = tfel::material::TangentOperatorTraits<tfel::material::MechanicalB                                                                                ehaviourBase::STANDARDSTRAINBASEDBEHAVIOUR>::SMFlag]'
PhaseFieldDisplacementNguyenSplit-generic.cxx:618:145:   required from here
PhaseFieldDisplacementNguyenSplit.mfront:21:32: error: 'using StiffnessTensor = tfel::config::Types<1, double, false>::StiffnessTensor' {                                                                                aka 'struct tfel::math::st2tost2<1, double>'} has no member named 'computeEigenVectors'
compilation terminated due to -Wfatal-errors.
make: *** [PhaseFieldDisplacementNguyenSplit-generic.o] Error 1
callMake: can't build target 'all'
libraries building went wrong
`
I am new in mfront so I propably did something wrong. I attached the .mfront file (in .txt format).

Thank you very much for the help.

Kind regards,

Lamia
[PhaseFieldDisplacementNguyenSplit.txt](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/9752173/PhaseFieldDisplacementNguyenSplit.txt)

**Comment by thelfer** [comment]
Hi Lamia, Unfortunately, there is no function to compute a eigenvalues of a fourth order tensor so far in MFront. But, in the orthotropic case, I think that we can use a small trick. Let me think of it.

**Reply by thelfer** [reply]
I think that the following code shall do the trick:

```cpp
namespace tfel::math{

template <unsigned short N, typename ValueType>
std::tuple<
    tvector<StensorDimeToSize<N>::value, ValueType>,
    tvector<StensorDimeToSize<N>::value, stensor<N, base_type<ValueType>>>>
diagonalize(const st2tost2<N, ValueType> &C) {
  using real = base_type<ValueType>;
  auto e = [](const unsigned short i) constexpr {
    auto s = stensor<N, real> (real(0));
    s[i] = 1;
    return s;
  };
  constexpr const auto cste = Cste<ValueType>::sqrt2;
  constexpr auto e1 = e(0);
  constexpr auto e2 = e(1);
  constexpr auto e3 = e(2);
  auto tmp =
      stensor<3u, ValueType>{C(0, 0),        C(1, 1),        C(2, 2),
                             cste * C(0, 1), cste * C(0, 2), cste * C(1, 2)};
  const auto [vp, m] =
      tmp.template computeEigenVectors<stensor_common::FSESJACOBIEIGENSOLVER>();
  auto ns = tvector<StensorDimeToSize<N>::value, stensor<N, real>>{};
  ns[0] = m(0, 0) * e1 + m(1, 0) * e2 + m(2, 0) * e3;
  ns[1] = m(0, 1) * e1 + m(1, 1) * e2 + m(2, 1) * e3;
  ns[2] = m(0, 2) * e1 + m(1, 2) * e2 + m(2, 2) * e3;
  if constexpr (N == 1) {
    return {vp, ns};
  } else   if constexpr (N == 2) {
    ns[3] = e(3);
    return {tvector<4u, ValueType>{vp(0), vp(1), vp(2), C(3, 3)}, ns};
  } else {
    ns[3] = e(3);
    ns[4] = e(4);
    ns[5] = e(5);
    return {
        tvector<6u, ValueType>{vp(0), vp(1), vp(2), C(3, 3), C(4, 4), C(5, 5)},
        ns};
  }
} // end of diagonalize

} // end of namespace tfel::math
```

It computes the eigenvalues and the eigen tensors of a fourth order tensor assuming an orthotropic symmetry.

The simpliest way to use it in `MFront` is to do the following:

```cpp
@Includes{
#ifndef TFEL_MATH_ST2TOST2_DIAGONALIZE 
#define TFEL_MATH_ST2TOST2_DIAGONALIZE 

namespace tfel::math{

template <unsigned short N, typename ValueType>
std::tuple<
    tvector<StensorDimeToSize<N>::value, ValueType>,
    tvector<StensorDimeToSize<N>::value, stensor<N, base_type<ValueType>>>>
diagonalize(const st2tost2<N, ValueType> &C) {
  using real = base_type<ValueType>;
  .....
} // end of diagonalize

} // end of namespace tfel::math

#endif
}
```

**Reply by lmersel** [reply]
Hello,
I am sorry for the late answer and thank you very much for your help ! The compilation worked perfectly. I struggled a little bit with the understanding of the square root of 2 on some Cij components, but it should ok now. I will try to add the rest of the law and post when I got some results.

Best, 

Lamia

**Reply by lmersel** [reply]
Hello, 

I am pursuing the work on the implementation of the fracture behavior law (Nguyen et al. 2020). I added two return variables Cp1_2 and Cn1_2 for each case of N, in the diagonalize function you proposed. They represent the positive and negative square root of the stiffness stensor C.
I also completed the @Integrator{} block. The purpose of the law is to project strain tensor in the eigenspace of the stiffness tensor C. So the law introduces a transformed strain that I called 'et', such as 'et = Cp1_2 * e'.
I need to compute positive and negative part of 'et' and the positive projector tensor 'Ptp' = detp_det. 
The stiffness degradation C(d) is computed as follows : 
![image](https://user-images.githubusercontent.com/110243126/199721656-8922faa9-2c28-4239-a769-512c19f7f132.png)
 
But I met an issue at the compilation due to the declaraton of the variables I need when I call the function diagonalize(). 
I am not a regular user of cpp, then sometimes I struggle to correct the mistakes. 
I joined an extract of the law.

Thank you very much for your help,

Best regards,

Lamia
[PhaseFieldDisplacementNguyenSplit.txt](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/9929910/PhaseFieldDisplacementNguyenSplit.txt)

**Reply by lmersel** [reply]
With some help, I finally updated the code in fixing some errors. 
[PhaseFieldDisplacementNguyenSplit.txt](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/9930640/PhaseFieldDisplacementNguyenSplit.txt)

The new issue is about the product between the tensor 'Cp1_2' and the strain, thats logical. 
`PhaseFieldDisplacementNguyenSplit.mfront:143:32: error: no match for 'operator*' (operand types are 'std::tuple_element<0, std::tuple<tfel::math::tmatrix<3, 3, tfel::math::stensor<1, double> >, tfel::math::tmatrix<3, 3, tfel::math::stensor<1, double> > > >::type' {aka 'tfel::math::tmatrix<3, 3, tfel::math::stensor<1, double> >'} and 'const StrainStensor' {aka 'const tfel::math::stensor<1, double>'})
compilation terminated due to -Wfatal-errors.
make: *** [PhaseFieldDisplacementNguyenSplit-generic.o] Error 1
` 
I keep fixing the code :)

Best,

Lamia

**Reply by thelfer** [reply]
Dear Lamia,

The returned value from the diagonalize function was wrong, causing the issue.
See this corrected file:  [PhaseFieldDisplacementNguyenSplit.zip](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/9930840/PhaseFieldDisplacementNguyenSplit.zip)

However, I am pretty sure that you made some mistake in the computation of the  square root (sqrt(x)) of the stiffness stensor C because you don't take the square root of the eigen values. 

Best,

Thomas

**Reply by lmersel** [reply]
Dear Thomas, 

Thank you for your answer. Indeed, you are right. I made some mistakes and corrected them. Thanks !
For the negative square root of the stiffness, I used power<-1,2>(x).
I have a small question in the calculation of the positive and the negative square root of C. For the case N=1, can i express the tensor like that 
`   if constexpr (N == 1) {
      // pos and neg square root of stiffness tensor C
      auto Cp1_2 = sqrt(vp(0)) * (ns[0] ^ ns[0]) ;
      auto Cn1_2 = 1/sqrt(vp(0)) * (ns[0] ^ ns[0]); 
      return {vp, ns, Cp1_2, Cn1_2};`
Do I need to add ` sqrt(vp(1)) * (ns[1] ^ ns[1]) + sqrt(vp(2))*(ns[2] ^ ns[2]);`, since it is 1D ?
Same question for the 2D case, since the case can be reduced like
![image](https://user-images.githubusercontent.com/110243126/199922287-40571b64-333b-4790-a7ec-fb06d377e4cd.png)
(in the kelvon basis E1 =e1^e1, E2 = e2^e2, E3 = (e1^e2 + e2 ^e1)/2
and his spetral decomposition 
![image](https://user-images.githubusercontent.com/110243126/199922799-da4ad507-cda2-486a-a23e-8f87e028e337.png)
(ai : the eigen vector)

Thank for your help,

Best,

Lamia

**Reply by thelfer** [reply]
> Do I need to add sqrt(vp(1)) * (ns[1] ^ ns[1]) + sqrt(vp(2))*(ns[2] ^ ns[2]);, since it is 1D ?

Yes. There are three eigenvalues.

To be honest, I am now pretty sure that this code may not work in 1D or 2D because there is no reason that the Kelvin modes respects the "space" symmetry, but the `^` operator forces this (for instance, the result of this operator is forced to be diagonal in 1D).

**Reply by lmersel** [reply]
Thank for your remark. Indeed I am agreed. I am thinking about setting aside the 1D case and focusing on 2D et 3D, since I will never work in 1D. Otherwise, I could compute separatly the 1D case. 

If  I use the plane strain assumption (N=2), in small strain, I should provide a vector epsilon of size 4x1( e = [e00, e11, e22, sqrt(2)*e01]^T ). Thus the transformed strain, 'et' that I want to compute is the product of et = C^(1/2) : e with C^(1/2) a 4x4 matrix. 
C = [ [C00,                      C01,            C02,          sqrt(2)*C03],
         [C01,                      C11,            C12,          sqrt(2)*C13],
         [C02,                      C12,            C22,          sqrt(2)*C23],
         [ sqrt(2)*C03,   sqrt(2)*C13,   sqrt(2)*C23,     2*C33]]   

In 2D or 3D, the operator '^' used shouldn't provide a diagonal matrix right?  
![image](https://user-images.githubusercontent.com/110243126/200000173-6f6de204-2090-4e12-b23e-63d8e158c514.png).
In the article I read, they use a generic formula called the Sylvester's formula to compute the eigenprojector wi^wi.

Thanks,

Lamia
